src/Scanner.js

```diff
@@ -246,12 +252,16 @@ class DefaultScanner {
 
             if (element.parentElement) {
                 element = element.parentElement;
-                const inputs = element.querySelectorAll(this.matching.cssSelector('formInputsSelector'));
-                const buttons = element.querySelectorAll(this.matching.cssSelector('submitButtonSelector'));
-                // If we find a button or another input, we assume that's our form
-                if (inputs.length > 1 || buttons.length) {
-                    // found related input, return common ancestor
-                    return element;
+                // check if current element is the only child of the parent, and increase traversal only then
+                if (element.childElementCount > 1) {
+                    const inputs = element.querySelectorAll(this.matching.cssSelector('formInputsSelector'));
+                    const buttons = element.querySelectorAll(this.matching.cssSelector('submitButtonSelector'));
+                    // If we find a button or another input, we assume that's our form
+                    if (inputs.length > 1 || buttons.length) {
+                        // found related input, return common ancestor
+                        return element;
+                    }
+                    traversalLayerCount++;
```

Comment:
Optimizing the traversal count here -> only doing so when the parent is a meaningful container (with more than one child, and itself not being a shadow root).

Reply:
Just playing devil's advocate. Is this now at risk of being an infinite loop? I don't think so, because we have the guard against the parent being present and the element not being the document. We should be good. Just give it a further thought.